src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStreamAsyncMemoryOverloads.cs

```diff
@@ -0,0 +1,148 @@
+                                                                                        s_localizableDescriptionRead,
+                                                                                        isPortedFxCopRule: false,
+                                                                                        isDataflowRule: false);
+
+        internal static DiagnosticDescriptor PreferStreamWriteAsyncMemoryOverloadsRule = DiagnosticDescriptorHelper.Create(
+                                                                                        RuleId,
+                                                                                        s_localizableTitleWrite,
+                                                                                        s_localizableMessageWrite,
+                                                                                        DiagnosticCategory.Performance,
+                                                                                        RuleLevel.IdeSuggestion,
+                                                                                        s_localizableDescriptionWrite,
+                                                                                        isPortedFxCopRule: false,
+                                                                                        isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+            ImmutableArray.Create(PreferStreamReadAsyncMemoryOverloadsRule, PreferStreamWriteAsyncMemoryOverloadsRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(AnalyzeCompilationStart);
+        }
+
+        private void AnalyzeCompilationStart(CompilationStartAnalysisContext context)
+        {
+            if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemMemory1, out _) ||
+                !context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemIOStream, out INamedTypeSymbol? streamType))
+            {
+                return;
+            }
+
+            context.RegisterOperationAction(context =>
+            {
+                IInvocationOperation invocation = (IInvocationOperation)context.Operation;
+                IMethodSymbol method = invocation.TargetMethod;
+                DiagnosticDescriptor rule;
+
+                if (string.Equals(method.Name, "ReadAsync", StringComparison.Ordinal))
+                {
+                    rule = PreferStreamReadAsyncMemoryOverloadsRule;
+                }
+                else if (string.Equals(method.Name, "WriteAsync", StringComparison.Ordinal))
+                {
+                    rule = PreferStreamWriteAsyncMemoryOverloadsRule;
+                }
```

Comment:
I don't think we should be using string-based comparisons against the method name like this... from my understanding, this is going to result in a fairly expensive operation, given that this is now being invoked for every invocation.  Further, if the new methods aren't available, there's no point in doing any checks.

So, this should be restructured akin to this:
```C#
private void AnalyzeCompilationStart(CompilationStartAnalysisContext context)
{
    // Try to get symbol for Stream; if you can't, return.
    // Try to get the symbols for the byte[]-based ReadAsync/WriteAsync on Stream; if you can't, return.
    // Try to get the symbols for the Memory-based ReadAsync/WriteAsync on Stream; if you can't, return.

    context.RegisterOperationAction(context =>
    {
        IInvocationOperation invocation = (IInvocationOperation)context.Operation;
        // If invocation.TargetMethod != one of the byte[]-based Read/WriteAsync symbols or an override of one of them, return.

        ...
    }, OperationKind.Invocation);
}
```

Reply:
OK I will restructure the code.

Reply:
Agreed - if you can identify the exact set of symbols to analyze upfront, that would be the most efficient implementation.

Reply:
Done.